### ISSUE TITLE
Improve I/O counter handling

### ISSUE THREAD

**Issue by thewonderidiot** [virtualagc/virtualagc#1055]
There's been desire recently to trigger 1201/1202 program alarms at realistic times in the NASSP Apollo 11 descent, and for them to happen for realistic reasons. This means we need to hit accurate processor load. The primary (and I think, only) thing holding us back right now is a full simulation of all of the input and output counters, and their cycle-stealing increment instructions.

This is going to be a joint project with NASSP, with @indy91 doing most of the heavy lifting on their side.

Internally, the timers already generate PINC or DINC sequences at appropriate times as according to the scaler, and incur the proper amount of stolen time. This logic will be expanded to include all output counters (GYROD, CDUXD, CDUYD, CDUZD, TRUND, SHAFTD, THRSTD, EMSD, OTLNK, ALT).

The input counters will need to start using the increment sequences as well, and we'll have to figure out how the phasing of them will work. A lot of them have logic that, for example, doesn't allow pulses through unless specific channel bits are set, or things like that, that will need to be captured.

It's my intention to fully emulate the behavior of the counter cells, so each pending increment request will be serviced according to its priority (counters with lower addresses have higher priorities).

### Open Questions
* Should we handle PIPA pre-counting logic? Counts come in from the PIPAs at a fixed frequency, clocked by the AGC, but instead of generating increment instructions directly like CDUs the pulses are filtered through a pre-counter. This block only lets PIPAs counts through if there have been more than "normal" in one direction. Under no acceleration, the PIPAs generate constant 3 pulses in one direction and then 3 in the other; the pre-counter generates no counter increments until at least 4 in the same direction are received. It also performs sanity logic making sure pulses are coming in at the right frequency and whatnot.

* Should the interfacing application interleave count requests with calls to agc_engine(), or should all of the input counters have fifos like CduFifo that work through overall combined requests at a predetermined rate?

* How and when do we notify users of changed output counters? Do we have separate "counters" that we effectively move counts into? For example, for THRSTD, every time a throttle change is incremented via DINC we could increment the separate counter the other way, so the requested throttle would slowly move from one spot to the other.

**Comment by rburkey2005** [comment]
I agree with your approach, to the extent I understand it right now :smile:, and think it's what should be pursued.  But after reading it, it occurred to me to wonder if there might be a short-term, quick-and-dirty change that would allow triggering 1201/1202 alarms at will, whilst waiting for your correct fix to become available as the long-term solution?  I wonder if the following mightn't work in the very short term?

It looks to me like 1201 alarms are initiated by the executive as

 `TC      BAILOUT`
 `OCT   1201`

and similarly for 1202 alarms.  Or in the Apollo 11 code, it's `BAILOUT1`.  But regardless, suppose that the code sequence just mentioned was inserted into an unused portion of unswitched erasable memory or fixed-fixed memory and then a jump was performed to that location?

What I'd envisage is two new "fictitious" input channels (162 and 161), one that inserted `TC BAILOUT ; OCT 1201` at a location indicated by the fictitious input channel and jumped to it, and another that did the same for `TC BAILOUT ; OCT 1202`.  Or for that matter, if there's any free space, it should always be close to address 7777, so a 2-word block near there could always be used rather than having the input channel select an address.

The idea is that using this fictitious input channel would immediately trigger a 1201 or 1202.  This would also have the advantage that the alarms could be triggered at the times the mission logs indicated they _were_ triggered originally, consistently, while your correct fix would be not be deterministic, I think, for the timing, the type, nor even the total number of alarms triggered.  

Thoughts?  Traps I've missed?

**Comment by rburkey2005** [comment]
I forgot to mention that `BAILOUT` and `BAILOUT1` are in fixed-fixed memory, which is why my idea might be feasible at all.

I discovered a couple of traps in my own idea, though nothing too major.

First, a bailout triggered by a fictitious input channel couldn't necessarily occur immediately.  It couldn't happen during an interrupt, for example, nor after an EXTEND or INDEX.  I guess it could only occur at the same times interrupts could occur.  So it would have to be buffered and then only operated on when conditions were right.

Second, yaAGC doesn't actually know the address of `BAILOUT` or `BAILOUT1` and therefore couldn't insert a `TC` for it into memory.  One thing we could do instead is give yaYUL a command-line switch, say --bailout=_BailoutAddress_, and then have it automatically insert something equivalent to the following into the rope:

 `SETLOC  7776`
 `TC      BailoutAddress`

This could also be done as a manual operation on the rope if modifying yaYUL was too much hassle.

We'd need only one fictitious input channel (162), and what it would do is to pass an alarm code rather than an address.  When yaAGC operated on channel 162, it would stick that alarm code at 7777 and then would jump to 7776.  Thus we'd not only cut down on the number of fictitious i/o channels, but could trigger _any_ bailout code (1104, 1201, 1202, 1203, 1211) ... not that that's of any real interest, I guess.  Finally, it would mean that NASSP would no longer have to have any specialized knowledge about addresses within the rope, and would only need to worry about the alarm code it wanted to trigger, and when it wanted to trigger it.

**Comment by thewonderidiot** [comment]
Well, one downside, as opposed to causing processor overload, is that this only captures the alarms and not the other effects. The alarms were caused by multiple SERVICERs stacking up, and the very last thing a SERVICER did was update the DSKY. The processor load was such that the SERVICERs were finishing their control tasks, but being preempted before they could update the DSKY. As a result, the DSKY would occasionally stop updating, or if accompanied by a BAILOUT, just go blank. With the last alarm (I think?) the DSKY was blank for something like 10 seconds, and Neil said in the post-flight he didn't think it was ever coming back.

We could try to script that, too, of course. If we're just trying to trigger 1201s and 1202s, though, I wonder if it wouldn't be easier to just go through and mark all of the core sets and VACs as "used" so the next attempt to schedule a new job would fail, rather than attempting to change program flow?

**Comment by rburkey2005** [comment]
I suppose.  I don't actually like putting in a bunch of ad hoc stuff applying to just one special case, so I was rationalizing this with the lie that it was a minimally-invasive "general purpose" feature that might be reusable later.  :smile:   Plus I was too ignorant to even know about the DSKY thing anyway.  But you're right that it wouldn't be too hard to manually edit playback scripts to remove any DSKY updates for a few seconds prior to a channel 162 appearing in the playback script.

I guess an important question is how necessary a quick-and-dirty fix is anyway?  Just because my interpretation of what you suggested about the counters was that it would take a long time to implement doesn't mean that I'm right about it.

**Comment by thewonderidiot** [comment]
To be honest, I didn't know about it until a couple of weeks ago either :smile: For some unknown reason, the Seattle Book Company shipped me a copy of Don's new book super early (it is extremely good, by the way), and Don talks about all of this stuff in the chapter about the 11 landing.

You're not wrong about this stuff likely taking a while. I should be able to knock out the yaAGC side without too much trouble, but integration with NASSP is going to be a bear, since peripherals will need to be (potentially heavily) modified and the whole timestep needs to be rearranged. But it was going to happen eventually, anyway, so it's not like this is an unplanned thing. :smile: 

This is sort of why I was heading in the direction of using State->ExtraDelay in the short term to simulate processor overload without actually doing any of the work. It'd be super easy to increment that in the AGC loop in NASSP, and we could tune it to produce exactly the right amount of overload at exactly the right time. At least, it should work in theory, heh.

When you say channel 162 appearing in the playback, do you mean that what you want to have happening is yaAGC actually running the descent in a standalone configuration, being fed in channel and counter updates as recorded by NASSP? I had been picturing a canned DSKY recording, with yaAGC no longer in the loop.

By the way, do you recall reading anything about how MIT measured TLOSS? It would be useful to be able to calculate that correctly.

**Comment by indy91** [comment]
I haven't actually tried flooding the LGC with RR CDU pulses every timestep (60 fps) yet. So all the theoretical side effects of doing this at 60fps instead of constantly at the speed the CDU is able to haven't actually occured yet. That is the first thing I would like to do to try and cause the 1201/1202 alarms. Just lots of PCDU/MCDU.

**Comment by thewonderidiot** [comment]
@indy91 I don't think PCDU/MCDU, as currently implemented, actually steal any time, so yaAGC will happily do infinitely many of them without affecting anything.

**Comment by rburkey2005** [comment]
I guess that it's an oversight that PCDU/MCDU don't use any CPU time.  I wish I could say from memory why that is, but I just looked at the code and it looks like it's from some alien planet.  :smile:

@thewonderidiot No, it's just a script for the DSKY alone, but it has (or can have) any/all of the i/o channel transactions, so even though the DSKY doesn't care about channel 162, channel 162 could still be made to appear in the script, so it would be obvious where the edits to remove stuff were needed.

**Comment by rburkey2005** [comment]
I've created a new branch, Fictitious-1201, with the following experimental code:

- A CLI switch for yaYUL, --bailout=_BailoutName_, which puts a `TC BailoutName` at address 07776.  In other words, it prepares an AGC rope for use with the fictitious input channel 162 I proposed earlier.
- Luminary099 is assembled using --bailout=BAILOUT1.
- Luminary099.binsource is similarly modified (05616 is put at 07776, which is normally @).
- All other AGC versions are left unchanged.
- yaAGC is modified to have a fictitious channel 162, for which: if 020000 is written to the channel, output messages channels 010 and 011 are subsequently blocked, so the DSKY never updates; this block is removed when anything else is written to the channel.  If 01104, 01201, 01202, 01203, or 01211 is written to the channel, this value is deposited in address 07777 in memory, and program control is diverted to address 07776, though this isn't done unless the conditions are also appropriate for an interrupt to occur; i.e., it doesn't happen during an interrupt, EXTEND, or INDEX, or various other conditions.  Thus, a bailout with one of those numbers as the program alarm occurs.
- A Python program, fakeAlarms.py, has been added to the piPeripheral directory.  This program performs a one-shot write to any channel of the AGC.  You can see the full options with "fakeAlarms.py --help", but the default channel it writes to is 162 and the default value it writes is 01201, so if used without arguments it triggers a 1201.  The following sequence would freeze a DSKY and then 10 seconds later send a 1201 bailout command to a running copy of yaAGC:

 fakeAlarms.py --value=20000
 sleep 10
 fakeAlarms.py

Similarly, the following sequence would trigger a 1202 alarm:

 fakeAlarms.py --value=1202

It seems to work, though in the limited fashion I'm capable of using it, it isn't too impressive.  At any rate, it's there to experiment with if anybody's interested.

**Comment by thewonderidiot** [comment]
Alright, I'm finally done filling in the table of counters above. That took longer than I expected -- every counter (or at least, peripheral) is a special snowflake. Anyways, now that I've captured timing information for all of them, I'll start implementing the priority chain and fleshing out the scaler in yaAGC.

**Comment by thewonderidiot** [comment]
Since the yaAGC logic is going to be grouped by sequence priority and scaler timepulses, I thought it would be useful to list out priorities and rearrange the above info into timepulse actions:

## AGC Operation Priorities
1. Advance the scaler, always
2. Finish the remaining MCTs of the previous instruction, if any.
3. Perform the GOJAM instruction _and_ the `TC 4000` it generates.
4. Perform any pending counter interrupts.
5. Perform the highest pending processor interrupt.
6. Perform the next instruction.

## Timepulses
| TP | Actions |
| --- | -------- |
| T05A | ... |

**Comment by thewonderidiot** [comment]
Actually, correction to the operation priority list above -- GOJAM instructions happen the MCT _immediately after_ a restart is triggered, regardless of whether or not the executing instruction has actually completed. That's enough to make sure that erasable memory gets written to, so there are no parity errors, but does not guarantee, for a few instructions (such as multiply and divide), that the instruction has run to completion. It doesn't matter for the majority of instructions, since most of them just have STD2 as a second MCT sequence, which exists only to fetch the next instruction (and handle EXTEND, INHINT, and RELINT if they are that next instruction).

**Comment by thewonderidiot** [comment]
Right, so there are three instructions where this matters: DAS, DXCH, and (ED)RUPT. All three of those can perform writes to erasable in their second or later MCT, which means a restart during any of them could end up with only half of the data written out.

**Comment by thewonderidiot** [comment]
@rburkey2005 I'm making solid progress reorganizing things in yaAGC to prepare for counters. It's much cleaner now, particularly the scaler logic, which is currently an arcane tangle of magical spaghetti. In organizing things such that MCTs are taken in order, according to the "AGC Operation Priorities" list above, I've been able to make some simplifications. In particular, ExtraDelay is no longer needed. I'm also shifting the instruction processing so that it happens in the first MCT instead of the last, since that is correct in almost all cases. So I think I don't need PendFlag, either.

I happily deleted both of them in my development code, but ran into errors because both are saved and loaded to/from the core dump files. Is there any desirement to try to retain backward compatibility for those? Should I keep them around simply for loading and saving zeroes?

**Comment by rburkey2005** [comment]
@thewonderidiot Well, of course it's hard with open source to know what people may or may not be doing with it.  The only thing that immediately occurs to _me_ is that I may have suggested to people like @indy91 that they could perform setups for given mission scenarios directly in yaAGC and then preserve the core dump files to reproduce those setups later.  I doubt that anyone has actually done that, but I don't really know.  My guess would be that saving zeroes in those positions is probably better, just in case.  Perhaps it may also be good to tweak the timing of when the core files are saved to insure that the files would only be saved at times when ExtraDelay and PendFlag _would_ have been zero were they actually still in use.  Then the core dump files would be guaranteed to be 100% compatible.

**Comment by thewonderidiot** [comment]
Well, I'm most of the way through implementing the first group of I/O counters, the RHC/BMAG axes, and found something interesting. After doing a lot of math, I've confirmed that the interface modules in the AGCs in LM-1, LM-2, and LM-3 (and presumably in CMCs of the era too, although the circuits were unused) were electrically different from modules flown on LM-4 and later. A couple of resistors (R3 and R7) in the A-type interface circuits were replaced to change the maximum number of counts in the channel from 32 to 42, given the same input voltage from the RHC. I'm not sure exactly when this happened, but I do know that Sundance expected 32, and Luminary 69 expects 42.

(Un)luckily for us, the only programs we have that uses the 32 number are Aurora 12 (in its very-early-development DAP) and Sunburst 37. The RHC code was deleted from Sunburst 120 since the RHCs themselves were deleted from LM-1 and LM-2.

So, I'm either going to hardcode the scale factor to produce 42 counts at max scale, or make it an option that defaults to 42, since 42 is correct for almost all of the ropes, and all of them that are flight-worthy.

**Comment by thewonderidiot** [comment]
The other day I had a brief discussion with @indy91 about what the best interface to counters for NASSP was. We haven't worked through the details of input counters like the CDUs and PIPAs, but for output counters, he thinks the best interface is a function similar to ChannelOutput(). This function (say, OutputPulse()) would take as its sole argument an identifier for the line being pulsed. The function's implementation would be up to VirtualAGC integrators, and we would provide a sample implementation in SocketAPI or similar (this implementation could possibly collate pulses and send them out over fictitious I/O channels as worked previously).

For example, for the altitude tape meter output, yaAGC logic would call
```
OutputPulse(OUTPUT_ALT_0);
OutputPulse(OUTPUT_ALT_1);
```
functions for 0s and 1s being shifted out, respectively. The implementation of OutputPulse, for these two signals, would perform any processing necessary to reconstruct values. For these altitude outputs, the function would shift 0s and 1s onto a running temporary variable, and as soon as bit 16 becomes a 1 (which is how the interface does syncing), the data would be passed along to the tape meters and the temporary variable zeroed for the next output word.

More standard counter outputs, like the CDU drives, would simply either pass along counts to their peripherals step by step, or the ChannelOutput() function would integrate the pulses for the total time of one frame, and then be passed along.

Niklas, does this sound right to you?

**Comment by rburkey2005** [comment]
> For example, for the altitude tape meter output, yaAGC logic would call
>
> |OutputPulse(OUTPUT_ALT_0); OutputPulse(OUTPUT_ALT_1); |
>
> ...
>
> More standard counter outputs, like the CDU drives, would simply 
> either pass along counts to their peripherals step by step, or the 
> ChannelOutput() function would integrate the pulses for the total time 
> of one frame, and then be passed along.
>
>
For what it's worth, sounds good to me.

**Comment by indy91** [comment]
Yeah, I can work with that. And it shouldn't make things much more complicated on the NASSP side either.
